### ISSUE TITLE
Windows: GID column consistency across user, groups, processes tables

### ISSUE THREAD

**Issue by Apptestingrd** [osquery/osquery#4264]
I asked that question in slack channel but I couldn't find an answer
it seems like an inconsistent situation

![users and processes table inconsistency](https://user-images.githubusercontent.com/38127874/38415682-f805d802-399a-11e8-8e73-2a7f49deaa4f.png)

**Comment by obelisk** [comment]
Can you tell us which version of osquery you're using?

**Comment by Apptestingrd** [comment]
2.11.2

**Comment by Smjert** [comment]
I'm not sure what's correct here but with latest master (22d614cba299f43491f73fd71b51d8caaec78c3a), if I run osqueryd in a powershell with Administrator privilege, I have:
```
osquery>  SELECT uid,gid from processes where path like "%osqueryd%";
+------+-----+
| uid  | gid |
+------+-----+
| 1001 | 513 |
+------+-----+
```
where 513 is `Domain Users` (https://docs.microsoft.com/en-us/troubleshoot/windows-server/identity/security-identifiers-in-windows).

Querying for the `users` table shows 513 again:
```
osquery> select * from users;
+------+-----+------------+------------+--------------------+-------------------------------------------------------------------------------------------------+---------------------------------------------+-----------------------------+-----------------------------------------------+---------+
| uid  | gid | uid_signed | gid_signed | username           | description                                                                                     | directory                                   | shell                       | uuid                                          | type    |
+------+-----+------------+------------+--------------------+-------------------------------------------------------------------------------------------------+---------------------------------------------+-----------------------------+-----------------------------------------------+---------+
| 500  | 513 | 500        | 513        | Administrator      | Built-in account for administering the computer/domain                                          |                                             | C:\Windows\System32\cmd.exe | S-1-5-21-2498975115-2405240309-114662596-500  | local   |
| 503  | 513 | 503        | 513        | DefaultAccount     | A user account managed by the system.                                                           |                                             | C:\Windows\System32\cmd.exe | S-1-5-21-2498975115-2405240309-114662596-503  | local   |
| 501  | 513 | 501        | 513        | Guest              | Built-in account for guest access to the computer/domain                                        |                                             | C:\Windows\System32\cmd.exe | S-1-5-21-2498975115-2405240309-114662596-501  | local   |
| 1001 | 513 | 1001       | 513        | smjert             |                                                                                                 | C:\Users\smjert                             | C:\Windows\System32\cmd.exe | S-1-5-21-2498975115-2405240309-114662596-1001 | local   |
| 1002 | 513 | 1002       | 513        | sshd               |                                                                                                 |                                             | C:\Windows\System32\cmd.exe | S-1-5-21-2498975115-2405240309-114662596-1002 | local   |
| 504  | 513 | 504        | 513        | WDAGUtilityAccount | A user account managed and used by the system for Windows Defender Application Guard scenarios. |                                             | C:\Windows\System32\cmd.exe | S-1-5-21-2498975115-2405240309-114662596-504  | local   |
| 18   | 18  | 18         | 18         | SYSTEM             |                                                                                                 | %systemroot%\system32\config\systemprofile  | C:\Windows\system32\cmd.exe | S-1-5-18                                      | special |
| 19   | 19  | 19         | 19         | LOCAL SERVICE      |                                                                                                 | %systemroot%\ServiceProfiles\LocalService   | C:\Windows\system32\cmd.exe | S-1-5-19                                      | special |
| 20   | 20  | 20         | 20         | NETWORK SERVICE    |                                                                                                 | %systemroot%\ServiceProfiles\NetworkService | C:\Windows\system32\cmd.exe | S-1-5-20                                      | special |
+------+-----+------------+------------+--------------------+-------------------------------------------------------------------------------------------------+---------------------------------------------+-----------------------------+-----------------------------------------------+---------+
```

But querying for the groups shows that Administrators is 544 and Users is 545:
```
osquery> select * from groups;
+-----+------------+-------------------------------------+--------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| gid | gid_signed | groupname                           | group_sid    | comment                                                                                                                                                                                                        |
+-----+------------+-------------------------------------+--------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| 579 | 579        | Access Control Assistance Operators | S-1-5-32-579 | Members of this group can remotely query authorization attributes and permissions for resources on this computer.                                                                                              |
| 544 | 544        | Administrators                      | S-1-5-32-544 | Administrators have complete and unrestricted access to the computer/domain                                                                                                                                    |
| 551 | 551        | Backup Operators                    | S-1-5-32-551 | Backup Operators can override security restrictions for the sole purpose of backing up or restoring files                                                                                                      |
| 569 | 569        | Cryptographic Operators             | S-1-5-32-569 | Members are authorized to perform cryptographic operations.                                                                                                                                                    |
| 583 | 583        | Device Owners                       | S-1-5-32-583 | Members of this group can change system-wide settings.                                                                                                                                                         |
| 562 | 562        | Distributed COM Users               | S-1-5-32-562 | Members are allowed to launch, activate and use Distributed COM objects on this machine.                                                                                                                       |
| 573 | 573        | Event Log Readers                   | S-1-5-32-573 | Members of this group can read event logs from local machine                                                                                                                                                   |
| 546 | 546        | Guests                              | S-1-5-32-546 | Guests have the same access as members of the Users group by default, except for the Guest account which is further restricted                                                                                 |
| 578 | 578        | Hyper-V Administrators              | S-1-5-32-578 | Members of this group have complete and unrestricted access to all features of Hyper-V.                                                                                                                        |
| 568 | 568        | IIS_IUSRS                           | S-1-5-32-568 | Built-in group used by Internet Information Services.                                                                                                                                                          |
| 556 | 556        | Network Configuration Operators     | S-1-5-32-556 | Members in this group can have some administrative privileges to manage configuration of networking features                                                                                                   |
| 559 | 559        | Performance Log Users               | S-1-5-32-559 | Members of this group may schedule logging of performance counters, enable trace providers, and collect event traces both locally and via remote access to this computer                                       |
| 558 | 558        | Performance Monitor Users           | S-1-5-32-558 | Members of this group can access performance counter data locally and remotely                                                                                                                                 |
| 547 | 547        | Power Users                         | S-1-5-32-547 | Power Users are included for backwards compatibility and possess limited administrative powers                                                                                                                 |
| 555 | 555        | Remote Desktop Users                | S-1-5-32-555 | Members in this group are granted the right to logon remotely                                                                                                                                                  |
| 580 | 580        | Remote Management Users             | S-1-5-32-580 | Members of this group can access WMI resources over management protocols (such as WS-Management via the Windows Remote Management service). This applies only to WMI namespaces that grant access to the user. |
| 552 | 552        | Replicator                          | S-1-5-32-552 | Supports file replication in a domain                                                                                                                                                                          |
| 581 | 581        | System Managed Accounts Group       | S-1-5-32-581 | Members of this group are managed by the system.                                                                                                                                                               |
| 545 | 545        | Users                               | S-1-5-32-545 | Users are prevented from making accidental or intentional system-wide changes and can run most applications                                                                                                    |
+-----+------------+-------------------------------------+--------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

So it seems there's a mix between Local and Domain groups.
In `processes` and `users` we use the `NetUserGetInfo` API to retrieve the `uid` and `gid`, and specifically we take the `gid` from the `usri3_primary_group_id`, but that is a Domain group, while instead I believe we should use `NetUserGetLocalGroups`. Running the MSDN example on my user returns the correct group name.
The `groups` table shows the Local groups correctly because it uses the `NetLocalGroupEnum` API.

**Comment by manurautela** [comment]
@Smjert I took a look at this issue. And still it seems to be present on the current master.
Based the information so far in the issue. I noticed that most of the places, where we use the api to fetch information for SID. We seemed to passing nullptr as server name, which implies local groups only.

After various tries. Here is what i found that could possibly fix this.

**Problem:**
We use **TOKEN_USER** which returns User SID here. 
[https://github.com/osquery/osquery/blob/dcef5d54c83bf80ca1cd1dd86eb810b71605cc44/osquery/tables/system/windows/processes.cpp#L390](url)

But later on we interpret this as **PTOKEN_OWNER** (even though internally both structures are same and contain PSID)
[https://github.com/osquery/osquery/blob/dcef5d54c83bf80ca1cd1dd86eb810b71605cc44/osquery/tables/system/windows/processes.cpp#L416](url)

But when querying token information based on this information class **GetTokenInformation** will return different SID.
[https://github.com/osquery/osquery/blob/dcef5d54c83bf80ca1cd1dd86eb810b71605cc44/osquery/tables/system/windows/processes.cpp#L396](url)

Ideally we should use **TOKEN_USER** to create **UID** and **TOKEN_OWNER** for **GID**.

I ended up printing the various groups(via **TOKEN_GROUPS** as well) the user belonged to initially and eventually realized that **TOKEN_OWNER** itself should suffice for our case to determine **GID**.


```diff
diff --git a/osquery/tables/system/windows/processes.cpp b/osquery/tables/system/windows/processes.cpp
index bc705cac..372b9950 100644
--- a/osquery/tables/system/windows/processes.cpp
+++ b/osquery/tables/system/windows/processes.cpp
@@ -388,6 +388,18 @@ void genProcessUserTokenInfo(HANDLE& proc, DynamicTableRowHolder& r) {
   /// Get the process UID and GID from its SID
   HANDLE tok = nullptr;
   std::vector<char> tok_user(sizeof(TOKEN_USER), 0x0);
+#if 1
+  /// We should probably get sid for the group and then use that to convert to proper
+  /// Group id, currently we use the same token for both sid conversion
+  std::vector<char> tok_group(sizeof(TOKEN_GROUPS), 0x0); // TOKEN_GROUPS
+  bool ret_grp = false;
+
+  std::vector<char> tok_primary(sizeof(TOKEN_PRIMARY_GROUP), 0x0); // TOKEN_PRIMARY_GROUP
+  bool ret_primary = false;
+
+  std::vector<char> tok_owner(sizeof(TOKEN_OWNER), 0x0); // TOKEN_OWNER
+  bool ret_owner = false;
+#endif
 
   auto ret = OpenProcessToken(proc, TOKEN_READ, &tok);
   long elevated = -1;
@@ -400,6 +412,65 @@ void genProcessUserTokenInfo(HANDLE& proc, DynamicTableRowHolder& r) {
           tok, TokenUser, tok_user.data(), tokOwnerBuffLen, &tokOwnerBuffLen);
     }
 
+
+#if 1
+    ret_grp = GetTokenInformation(tok, TokenGroups, nullptr, 0, &tokOwnerBuffLen);
+    if (ret_grp == 0 && GetLastError() == ERROR_INSUFFICIENT_BUFFER) {
+      tok_group.resize(tokOwnerBuffLen);
+      ret_grp = GetTokenInformation(
+          tok, TokenGroups, tok_group.data(), tokOwnerBuffLen, &tokOwnerBuffLen);
+    }
+
+    // https://docs.microsoft.com/en-us/windows/win32/secauthz/searching-for-a-sid-in-an-access-token-in-c--
+    if (ret_grp != 0 && !tok_group.empty()) {
+        std::string gid_list = "";
+        PTOKEN_GROUPS grp_info = (PTOKEN_GROUPS)tok_group.data();
+        for(size_t i=0; i < grp_info->GroupCount; i++) {
+            auto gid = getGidFromSid(grp_info->Groups[i].Sid);
+            std::string gid_str = std::to_string(gid);
+            gid_list += gid_str;
+            gid_list += ",";
+        }
+        printf("\n\n**************************************************************\n");
+        printf("[+] List of groups based on TOKEN_GROUPS: %s\n", gid_list.c_str());
+    }
+
+#endif
+
+#if 0
+    ret_primary = GetTokenInformation(tok, TokenPrimaryGroup, nullptr, 0, &tokOwnerBuffLen);
+    if (ret_primary== 0 && GetLastError() == ERROR_INSUFFICIENT_BUFFER) {
+      tok_primary.resize(tokOwnerBuffLen);
+      ret_primary= GetTokenInformation(
+          tok, TokenPrimaryGroup, tok_primary.data(), tokOwnerBuffLen, &tokOwnerBuffLen);
+    }
+
+    if (ret_primary != 0 && !tok_primary.empty()) {
+        PTOKEN_PRIMARY_GROUP tok_primary_sid = (PTOKEN_PRIMARY_GROUP)tok_primary.data();
+        auto gid = getGidFromSid(tok_primary_sid->PrimaryGroup);
+        printf("Correct gid:%d based on PRIMARY_TOKEN_GROUP\n", gid);
+    }
+#endif
+
+
+#if 1
+    ret_owner = GetTokenInformation(tok, TokenOwner, nullptr, 0, &tokOwnerBuffLen);
+    if (ret_owner == 0 && GetLastError() == ERROR_INSUFFICIENT_BUFFER) {
+      tok_owner.resize(tokOwnerBuffLen);
+      ret_owner= GetTokenInformation(
+          tok, TokenOwner, tok_owner.data(), tokOwnerBuffLen, &tokOwnerBuffLen);
+    }
+
+    if (ret_owner != 0 && !tok_owner.empty()) {
+        PTOKEN_OWNER tok_owner_sid = (PTOKEN_OWNER)tok_owner.data();
+        auto gid = getGidFromSid(tok_owner_sid->Owner);
+        printf("[+] Correct gid:%d based on TOKEN_OWNER\n", gid);
+    }
+#endif
+
     /// Check if the process is using an elevated token
     TOKEN_ELEVATION elevation;
     DWORD cb_size = sizeof(TOKEN_ELEVATION);
@@ -413,11 +484,21 @@ void genProcessUserTokenInfo(HANDLE& proc, DynamicTableRowHolder& r) {
   }
   r["is_elevated_token"] = INTEGER(elevated);
   if (ret != 0 && !tok_user.empty()) {
+    // FIXME:
+    // BUILTIN\Administrators (TOKEN_OWNER), whereas TOKEN_USER is the
+    // "computer\user" PTOKEN_OWNER and PTOKEN_USER are both same struct but
+    // use diff information class used to fetch the sid are diff. In the
+    // current impl we pass TokenUser while getting token information and
+    // interpret as if it's TokenOwner which might actually be different.
+    // Hence the issue when reporting GID based off that.
     auto sid = PTOKEN_OWNER(tok_user.data())->Owner;
 
     r["uid"] = BIGINT(getUidFromSid(sid));
     r["gid"] = BIGINT(getGidFromSid(sid));
+
+    printf("[*] Currently fetched uid:%s gid:%s\n", r["uid"].c_str(), r["gid"].c_str());
   }
+
   if (tok != nullptr) {
     CloseHandle(tok);
   }
@@ -600,6 +681,7 @@ TableRows genProcesses(QueryContext& context) {
 
     if (context.isAnyColumnUsed({"uid", "gid", "is_elevated_token"})) {
       genProcessUserTokenInfo(proc_handle, r);
+      printf("[*] PID:%d Image:%ws\n", pid, proc.szExeFile);
     }
 
     if (context.isAnyColumnUsed({"user_time",
```




![image](https://user-images.githubusercontent.com/1333906/109826192-8ed59a00-7c60-11eb-8a8d-b361e0bcb70e.png)

**Comment by mike-myers-tob** [comment]
Ok, this is a complicated issue, but I think the solution was already identified by @Smjert. Still, I want to recap this here.

On Windows, users belong to multiple groups. The only correct `GID` column on Windows would be a list of GID values (as @manurautela demonstrated above), not a single GID like on POSIX. Since osquery wants to have a single schema for these tables across platforms, it must make a choice as to which group's `GID` to display on Windows. Maybe that's okay, but we want consistency between `processes` and the `users` and `groups` and `user_groups` tables. Currently what we have is inconsistent, confusing, and incompatible with `JOIN`s between these tables.

Currently, `processes` table presents a valid `GID` (`513` is a "well-known GID" as @smjert noted above), but not one that corresponds to any group listed in the `groups` table, because the `groups` table only shows "Local" groups, like you would see with a `net localgroup` command or `Get-LocalGroup | Select *` in PowerShell, like @Smjert said.

@manurautela has found something very interesting with his code above, but I am not sure we need to make this specific change. I'll explain.

When `osqueryd` runs in a standard Powershell, `TOKEN_OWNER` and `TOKEN_USER` give the same values. When `osqueryd` is run in an Administrator Powershell, they are different in the way that @manurautela noted. The 'user' is the same in both cases but apparently the impersonation (if that's the right term) changes the 'owner' GID.

Running this query with @manurautela 's code:

`SELECT uid, gid from processes where path like "%osquery%";`

When osqueryd is running in a standard Powershell:

```
**************************************************************
[+] List of groups based on TOKEN_GROUPS: 513,0,114,544,545,559,4,1,11,15,113,177197,0,10,8192,
[+] Correct gid:513 based on TOKEN_OWNER
[*] Currently fetched uid:1002 gid:513
+------+-----+
| uid  | gid |
+------+-----+
| 1002 | 513 |
+------+-----+
```

When osqueryd is running in an Administrator Powershell:

```
**************************************************************
[+] List of groups based on TOKEN_GROUPS: 513,0,114,544,545,559,4,1,11,15,113,177197,0,10,12288,
[+] Correct gid:544 based on TOKEN_OWNER
[*] Currently fetched uid:1002 gid:513
+------+-----+
| uid  | gid |
+------+-----+
| 1002 | 513 |
+------+-----+
```

Both processes belong to my user, `mmyers` (can confirm this in Task Manager). If we think the `gid` should change when a process is elevated, okay, we could make that change. But we already have the `is_elevated_token` column. 

## TLDR 

The problem:

1. the `groups` table doesn't list all the groups, only the "Local" ones. FWIW I think this was intentional because we didn't want it trying to enumerate Active Directory groups.
2. on Windows, a user is a member of multiple groups, and the first one in the list for the SID happens to be GID 513, which is not a local group.
3. in `getGidFromSid` (process_ops.cpp), called by the `users` and `processes` tables, we use `NetUserGetInfo` when maybe we should use `NetUserGetLocalGroups`

The solution:

1. leave `groups` table as-is, but change `getGidFromSid` to use `NetUserGetLocalGroups` or whatever fix is necessary for it to return the most sensible choice of "Local" group's GID.

**Comment by manurautela** [comment]
Thanks @mike-myers-tob for the info. It does makes sense to switch to local versions.  Good to see it getting fixed.
